### ISSUE TITLE
Support "symmetric" reflection padding

### ISSUE THREAD

**Issue by pyscorcher** [pytorch/pytorch#46240]
## 🚀 Feature

Currently `ReflectionPadNd` does *not* repeat the boundary pixels, as demonstrated by the docs:

```
>>> m = nn.ReflectionPad2d(2)
>>> input = torch.arange(9, dtype=torch.float).reshape(1, 1, 3, 3)
>>> input
tensor([[[[0., 1., 2.],
          [3., 4., 5.],
          [6., 7., 8.]]]])
>>> m(input)
tensor([[[[8., 7., 6., 7., 8., 7., 6.],
          [5., 4., 3., 4., 5., 4., 3.],
          [2., 1., 0., 1., 2., 1., 0.],
          [5., 4., 3., 4., 5., 4., 3.],
          [8., 7., 6., 7., 8., 7., 6.],
          [5., 4., 3., 4., 5., 4., 3.],
          [2., 1., 0., 1., 2., 1., 0.]]]])
```

According to the paper mentioned below, it would be very beneficial to use the `SYMMETRIC` variation that is already available in [tensorflow](https://www.tensorflow.org/api_docs/python/tf/pad) - I'm quoting the example from the linked page:

```
t = tf.constant([[1, 2, 3], [4, 5, 6]])
paddings = tf.constant([[1, 1,], [2, 2]])
# 'constant_values' is 0.
# rank of 't' is 2.
tf.pad(t, paddings, "CONSTANT")  # [[0, 0, 0, 0, 0, 0, 0],
                                 #  [0, 0, 1, 2, 3, 0, 0],
                                 #  [0, 0, 4, 5, 6, 0, 0],
                                 #  [0, 0, 0, 0, 0, 0, 0]]

tf.pad(t, paddings, "REFLECT")  # [[6, 5, 4, 5, 6, 5, 4],
                                #  [3, 2, 1, 2, 3, 2, 1],
                                #  [6, 5, 4, 5, 6, 5, 4],
                                #  [3, 2, 1, 2, 3, 2, 1]]

tf.pad(t, paddings, "SYMMETRIC")  # [[2, 1, 1, 2, 3, 3, 2],
                                  #  [2, 1, 1, 2, 3, 3, 2],
                                  #  [5, 4, 4, 5, 6, 6, 5],
                                  #  [5, 4, 4, 5, 6, 6, 5]]

```

I propose two ways in which this could be done: Either we extend `ReflectionPadXd()` with a `mode` argument, or alternatively we introduce `SymmetricPadXd()` as a separate class in `torch.nn`.

## Motivation

This is prompted by the paper [Mind the Pad -- CNNs can Develop Blind Spots](https://arxiv.org/abs/2010.02178v1) (Alsallakh et al, facebook AI), which investigated the behaviour of different padding methods and how some of them have a detrimental effect to the network.

## Pitch

Add a new method that pads an array `[1,2,3,4,5,6]` symmetically by repeating the boundary pixel  as `[2,1,1,2,3,4,5,6,6,5]` (unlike `ReflectionPadNd()` that does *not* repeat the boundary pixel: `[3,2,1,2,3,4,5,6,5,4]`).

## Alternatives

This behaviour can still be done using the current methods by first using a 1-pixel `ReplicationPadXd()` and add the `ReflectionPadXd()` after that, but it is quite cumbersome.

## Additional context

The *"symmetric padding"* is also popular in wavelet transforms or discrete convolutions, see e.g. [here](https://pywavelets.readthedocs.io/en/latest/ref/signal-extension-modes.html).

**Comment by mruberry** [comment]
Does NumPy's pad function's symmetric padding mode implement this behavior? (See https://numpy.org/doc/1.18/reference/generated/numpy.pad.html?highlight=pad#numpy.pad.)

I'd like us to consider implementing a pad function like that to simplify our padding UX.

**Comment by pyscorcher** [comment]
@mruberry good point! Yes, numpy behaves exactly like that:

```
>>> a = np.array([[0,1,2],[3,4,5],[6,7,8]])
>>> a
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])
>>> np.pad(a, 2, mode='symmetric')
array([[4, 3, 3, 4, 5, 5, 4],
       [1, 0, 0, 1, 2, 2, 1],
       [1, 0, 0, 1, 2, 2, 1],
       [4, 3, 3, 4, 5, 5, 4],
       [7, 6, 6, 7, 8, 8, 7],
       [7, 6, 6, 7, 8, 8, 7],
       [4, 3, 3, 4, 5, 5, 4]])
```

**Comment by mruberry** [comment]
Thank you for the follow-up, @pyscorcher. Looks like it's time to prioritize a pad implementation.

**Comment by rgommers** [comment]
There may be some synergy with or lessons from gh-45667 (`'same'` padding for conv1/2/3d). Cc @peterbell10

**Comment by peterbell10** [comment]
Most of the challenges I'm having with #45667 are codegen related. However, `F.pad` has separate manually created C++ and python APIs so I think sidesteps these issues entirely.

**Comment by avijain3** [comment]
Hi @pyscorcher, 
I have been trying to create a new class  `_SymmetricPadNd()`  in the `padding.py` file.
I am trying to understand the code from other classes like `_ConstantPadNd()` and `_ReplicationPadNd` that are already built but I encounter terms like 'constant' and 'replicate' in the `F.pad` function. I am unable to understand what exactly is their significance and are they imported keywords from somewhere?
Also, I saw the [https://pytorch.org/cppdocs/api/file_torch_csrc_api_include_torch_nn_functional_padding.h.html#file-torch-csrc-api-include-torch-nn-functional-padding-h](url) documentation and saw that it is defined in the `padding.h` header file.
So in order to create a `symmetric` type argument, would I have to change/add some code to `padding.h`?
I am fairly new and I apologize if this is a very naive question.
Thanks.

**Comment by underdogliu** [comment]
Hi all! So looks like I am facing similar issue as well on no symmetric padding. Is there any update from other relevant PRs this issue, if any?

**Comment by mruberry** [comment]
Not at this time, sorry @underdogliu. We would accept a PR implementing "symmetric" padding, compatible with that performed by NumPy's pad function, to PyTorch's existing torch.nn.functional.pad.

**Comment by underdogliu** [comment]
@mruberry ok I will see what I can do if have some time.

**Comment by Enmartz** [comment]
Is there some news about this support function?

**Comment by Yatagarasu50469** [comment]
Would very much appreciate this feature being added; currently using torchvision implementation as a workaround:
```
from torchvision.transforms import v2
v2.Pad(padding=1, padding_mode='symmetric')
```
As noted in the originally referenced paper, Replication padding is equivalent to symmetric if padding is 1, but it's (to my mind inexplicably) painfully expensive/slow...

**Comment by v0lta** [comment]
I have written a symmetric padding function in C++ for Pytorch tensors at https://github.com/v0lta/sympad_pytorch. The motivation is to have fast symmetric padding for the wavelet transforms at https://github.com/v0lta/PyTorch-Wavelet-Toolbox. Symmetric padding is the default at https://pywavelets.readthedocs.io/en/latest/. I think it's useful if we allow people to switch back and forth with ease. Consequently, we would like to have symmetric padding as the default in both libraries.

To extend functional.pad I think the code should go into https://github.com/pytorch/pytorch/blob/v1.13.1/aten/src/ATen/native/PadNd.cpp . But I haven't found a good place for the tests. Where should these go?

**Comment by v0lta** [comment]
I think I found it: https://github.com/pytorch/pytorch/blob/217aac96d779841666527402fc113493b6bd6323/test/test_nn.py#L8447

**Comment by v0lta** [comment]
I have manually edited  ATen/PadNd.h and opened a pull request.
